crates/sui/src/client_ptb/ast.rs

```diff
@@ -0,0 +1,353 @@
+pub const SPLIT_COINS: &str = "split-coins";
+pub const MERGE_COINS: &str = "merge-coins";
+pub const MAKE_MOVE_VEC: &str = "make-move-vec";
+pub const MOVE_CALL: &str = "move-call";
+pub const PUBLISH: &str = "publish";
+pub const UPGRADE: &str = "upgrade";
+pub const ASSIGN: &str = "assign";
+pub const PREVIEW: &str = "preview";
+pub const WARN_SHADOWS: &str = "warn-shadows";
+pub const GAS_BUDGET: &str = "gas-budget";
+pub const SUMMARY: &str = "summary";
+pub const GAS_COIN: &str = "gas-coin";
+pub const JSON: &str = "json";
+
+// Types
+pub const U8: &str = "u8";
+pub const U16: &str = "u16";
+pub const U32: &str = "u32";
+pub const U64: &str = "u64";
+pub const U128: &str = "u128";
+pub const U256: &str = "u256";
+
+// Keywords
+pub const ADDRESS: &str = "address";
+pub const BOOL: &str = "bool";
+pub const VECTOR: &str = "vector";
+pub const SOME: &str = "some";
+pub const NONE: &str = "none";
+pub const GAS: &str = "gas";
+
+pub const KEYWORDS: &[&str] = &[
+    ADDRESS, BOOL, VECTOR, SOME, NONE, GAS, U8, U16, U32, U64, U128, U256,
+];
+
+pub fn is_keyword(s: &str) -> bool {
+    KEYWORDS.contains(&s)
+}
+
+pub fn all_keywords() -> String {
+    KEYWORDS[..KEYWORDS.len() - 1]
+        .iter()
+        .map(|x| format!("'{}'", x))
+        .collect::<Vec<_>>()
+        .join(", ")
+        + &format!(", or '{}'", KEYWORDS[KEYWORDS.len() - 1])
+}
+
+/// A PTB Program consisting of a list of commands and a flag indicating if the preview
+/// warn-shadows command was present.
+#[derive(Debug, Clone)]
+pub struct Program {
+    pub commands: Vec<Spanned<ParsedPTBCommand>>,
+    // Held outside of metadata since this is used by the PTB builder
+    pub warn_shadows_set: bool,
+}
+
+/// The `ProgramMetadata` struct holds metadata about a PTB program, such as whether the preview
+/// flag was set, json output was set, etc.
+#[derive(Debug, Clone)]
+pub struct ProgramMetadata {
+    pub preview_set: bool,
+    pub summary_set: bool,
+    pub gas_object_id: Option<Spanned<ObjectID>>,
+    pub json_set: bool,
+    pub gas_budget: Spanned<u64>,
+}
+
+/// A parsed module access consisting of the address, module name, and function name.
+#[derive(Debug, Clone)]
+pub struct ModuleAccess {
+    pub address: Spanned<ParsedAddress>,
+    pub module_name: Spanned<Identifier>,
+    pub function_name: Spanned<Identifier>,
+}
+
+/// A parsed PTB command consisting of the command and the parsed arguments to the command.
+#[derive(Debug, Clone)]
+pub enum ParsedPTBCommand {
+    TransferObjects(Spanned<Argument>, Spanned<Vec<Spanned<Argument>>>),
+    SplitCoins(Spanned<Argument>, Spanned<Vec<Spanned<Argument>>>),
+    MergeCoins(Spanned<Argument>, Spanned<Vec<Spanned<Argument>>>),
+    MakeMoveVec(Spanned<ParsedType>, Spanned<Vec<Spanned<Argument>>>),
+    MoveCall(
+        Spanned<ModuleAccess>,
+        Option<Spanned<Vec<ParsedType>>>,
+        Vec<Spanned<Argument>>,
+    ),
+    Assign(Spanned<String>, Option<Spanned<Argument>>),
```

Comment:
I'd like to keep them in this order (roughly) if possible since there's a semantic ordering (at least in mind) to these since:
* `TransferObject`, `SplitCoins`, `MergeCoins` all take the same argument structure and are coin-related
* `MakeMoveVec` and `MoveCall` both take type arguments
* `Assign` is kinda on its own, so that can get moved up to the top :) (although actually as I think about it it should be grouped with `Preview` and `WarnShadows` I think? since it's a "meta command" that won't show up in the output)
* `Publish`/`Upgrade` are all package related commands
* `Preview` and `WarnShadows` are directives that won't show up in the output.

Reply:
Ok (me and semantic ordering have never got along :p)